### ISSUE TITLE
Consider polish `TableIOContext` further

### ISSUE THREAD

**Issue by dantengsky** [datafuselabs/databend#2168]
**Summary**

- Currently, `TableIOContext`, instead of `DatabendQueryContext` is passed to methods of trait `Table`

  I do appreciate the efforts of trying to "narrow the dependency" by using a more specific data structure when possible (hope I get this design decision right), but

in `DatabendQueryContext`, we construct `TableIOContext` like this

```
TableIOContext::new(
            self.get_shared_runtime()?,
            Arc::new(ContextDalBuilder::new(self.get_config().storage)),
            max_threads,
            nodes,
            Some(self.clone()),  // where self::Arc<DatabendQueryContext>
```
latter in implementation of Table, we access `DatabendQueryContext` like this

```
        let ctx: Arc<DatabendQueryContext> = io_ctx
            .get_user_data()?
            .expect("DatabendQueryContext should not be None");
```

This may be too much, deserve to be polished further.

- maybe we could impl `IOContext` for `DatabendQueryContext`, or extract methods of `DatabendQueryContext` into a trait that is dedicated to trait `Table`

see also
- issue #2135

**Comment by dantengsky** [comment]
\cc @BohuTANG @drmingdrmer

**Comment by BohuTANG** [comment]
I agree.
It should be:
DatabendQueryContext only contains others *Context

**Comment by drmingdrmer** [comment]
> * maybe we could impl `IOContext` for `DatabendQueryContext`, or extract methods of `DatabendQueryContext` into a trait that is dedicated to trait `Table`

I did not try every approach to solve the current issue that blocking refactoring the code base.

AFAIK, 
- to get DatabendQueryContext off `trait Table`, a new concrete type xxxContext is inevitable.
- Otherwise, you have to introduce a trait that contains every method a table would use.
  Then that trait has to depend on some types defined in `query`.

**Comment by dantengsky** [comment]
roger

**Comment by dantengsky** [comment]
being closed since addressed in PR #2879
